hastory-data-gen/src/Data/Hastory/Gen.hs

```diff
@@ -30,9 +30,7 @@ instance GenValid UserForm where
   shrinkValid = shrinkValidStructurallyWithoutExtraFiltering
 
 instance GenValid Username where
-  genValid =
-    mkUsername <$>
-    (genValid `suchThat` \rawUsername -> notNull rawUsername && T.all isAlphaNum rawUsername)
+  genValid = Username <$> (genValid `suchThat` \text -> notNull text && T.all isAlphaNum text)
     where
       notNull = not . T.null
-  shrinkValid = map mkUsername . filter (not . T.null) . shrinkValid . rawUserName
+  shrinkValid = map Username . filter (not . T.null) . shrinkValid . usernameText
```

Comment:
This is also wrong because it can shrink to invalid usernames.

Reply:
Yes - this code is incorrect. I have looked at `smos` and there seems to be a `UsernameChar` newtype wrapper around a `Char`; by creating instances of `GenValid` and `Validity` for this newtype, it seems we can then start using the `shrinkValidStructurally` and other helpers. I'll go down that route tomorrow.

Reply:
This code no longer exists.